File: internal/client/desktop.go
package client

import (
	"os"
	"runtime"

	"rdev/internal/protocol"
)

func desktopSourcesByBackend(backend string) []protocol.DesktopSource {
	return desktopSourcesByBackendFrom(desktopSources(), backend)
}

func desktopSourcesByBackendFrom(sources []protocol.DesktopSource, backend string) []protocol.DesktopSource {
	var filtered []protocol.DesktopSource
	for _, source := range sources {
		if source.ID == "auto" {
			continue
		}
		if source.Backend == backend {
			filtered = append(filtered, source)
		}
	}
	return filtered
}

func desktopCapabilities() *protocol.DesktopCapabilities {
	caps := &protocol.DesktopCapabilities{
		Platform:  runtime.GOOS,
		Supported: false,
		ViewOnly:  false,
		Input:     false,
		Clipboard: false,
	}

	switch runtime.GOOS {
	case "windows":
		sources := desktopSources()
		inputOptions := desktopInputOptions()
		inputBackends := desktopInputBackendIDs(inputOptions)
		caps.DisplayServer = "windows"
		caps.Supported = true
		caps.ViewOnly = false
		caps.Input = len(inputBackends) > 0
		caps.InputBackends = inputBackends
		caps.InputOptions = inputOptions
		caps.Backends = []string{"win32-gdi"}
		if len(desktopSourcesByBackendFrom(sources, "dxgi")) > 0 {
			caps.Backends = append(caps.Backends, "dxgi")
		}
		caps.Sources = sources
	case "linux":
		sources := desktopSources()
		inputOptions := desktopInputOptions()
		inputBackends := desktopInputBackendIDs(inputOptions)
		fbSources := desktopSourcesByBackendFrom(sources, "fbdev")
		drmSources := desktopSourcesByBackendFrom(sources, "drm-kms")
		x11Sources := append(desktopSourcesByBackendFrom(sources, "x11"), desktopSourcesByBackendFrom(sources, "x11-randr")...)
		if len(x11Sources) > 0 {
			caps.DisplayServer = "x11"
			caps.Supported = true
			caps.ViewOnly = false
			caps.Input = len(inputBackends) > 0
			caps.InputBackends = inputBackends
			caps.InputOptions = inputOptions
			caps.Backends = []string{"x11"}
			if len(drmSources) > 0 {
				caps.Backends = append(caps.Backends, "drm-kms")
			}
			caps.Sources = sources
		} else if os.Getenv("WAYLAND_DISPLAY") != "" {
			caps.DisplayServer = "wayland"
			caps.Input = len(inputBackends) > 0
			caps.InputBackends = inputBackends
			caps.InputOptions = inputOptions
			caps.Backends = []string{"wayland-portal"}
			caps.Reason = "Wayland capture requires native xdg-desktop-portal/PipeWire support; external tools are not used"
			if len(drmSources) > 0 {
				caps.Supported = true
				caps.ViewOnly = !caps.Input
				caps.Backends = append(caps.Backends, "drm-kms")
				caps.Sources = append([]protocol.DesktopSource{{ID: "auto", Label: "Auto", Kind: "screen", Backend: "drm-kms", Width: drmSources[0].Width, Height: drmSources[0].Height, Primary: true}}, append(drmSources, fbSources...)...)
				caps.Reason = "Wayland portal backend is not implemented; using DRM/KMS scanout fallback"
			} else if len(fbSources) > 0 {
				caps.Supported = true
				caps.ViewOnly = !caps.Input
				caps.Backends = append(caps.Backends, "fbdev")
				caps.Sources = append([]protocol.DesktopSource{{ID: "auto", Label: "Auto", Kind: "screen", Backend: "fbdev", Width: fbSources[0].Width, Height: fbSources[0].Height, Primary: true}}, fbSources...)
				caps.Reason = "Wayland portal backend is not implemented; using root framebuffer fallback"
			}
		} else if len(drmSources) > 0 {
			caps.DisplayServer = "drm-kms"
			caps.Supported = true
			caps.Input = len(inputBackends) > 0
			caps.ViewOnly = !caps.Input
			caps.InputBackends = inputBackends
			caps.InputOptions = inputOptions
			caps.Backends = []string{"drm-kms"}
			if len(fbSources) > 0 {
				caps.Backends = append(caps.Backends, "fbdev")
			}
			caps.Sources = append([]protocol.DesktopSource{{ID: "auto", Label: "Auto", Kind: "screen", Backend: "drm-kms", Width: drmSources[0].Width, Height: drmSources[0].Height, Primary: true}}, append(drmSources, fbSources...)...)
		} else if len(fbSources) > 0 {
			caps.DisplayServer = "fbdev"
			caps.Supported = true
			caps.Input = len(inputBackends) > 0
			caps.ViewOnly = !caps.Input
			caps.InputBackends = inputBackends
			caps.InputOptions = inputOptions
			caps.Backends = []string{"fbdev"}
			caps.Sources = append([]protocol.DesktopSource{{ID: "auto", Label: "Auto", Kind: "screen", Backend: "fbdev", Width: fbSources[0].Width, Height: fbSources[0].Height, Primary: true}}, fbSources...)
		} else {
			caps.Reason = "no desktop display or readable framebuffer detected"
		}
	case "darwin":
		inputOptions := desktopInputOptions()
		inputBackends := desktopInputBackendIDs(inputOptions)
		caps.DisplayServer = "quartz"
		caps.Supported = true
		caps.Input = len(inputBackends) > 0
		caps.ViewOnly = !caps.Input
		caps.InputBackends = inputBackends
		caps.InputOptions = inputOptions
		caps.Backends = []string{"quartz"}
		caps.Sources = desktopSources()
		caps.Reason = "macOS Quartz capture requires Screen Recording permission; input requires Accessibility permission"
	default:
		caps.Reason = "unsupported platform"
	}

	return caps
}

File: clients/rdev-client-gpu/src/desktop.rs
use crate::protocol::DesktopCapabilities;
#[cfg(feature = "embedded-rdev-desktop")]
use crate::protocol::DesktopSource;

pub fn capabilities(enabled: bool) -> DesktopCapabilities {
    if !enabled {
        return DesktopCapabilities {
            platform: std::env::consts::OS.to_string(),
            supported: false,
            view_only: true,
            input: false,
            clipboard: false,
            reason: "embedded GPU desktop service is not active".to_string(),
            ..Default::default()
        };
    }

    #[cfg(feature = "embedded-rdev-desktop")]
    {
        return DesktopCapabilities {
            platform: std::env::consts::OS.to_string(),
            display_server: display_server(),
            supported: true,
            view_only: false,
            input: true,
            clipboard: false,
            backends: desktop_backends(),
            input_backends: vec!["rdev-desktop".to_string()],
            video_codecs: vec!["h264".to_string()],
            encoder_backends: encoder_backends(),
            sources: vec![DesktopSource {
                id: "auto".to_string(),
                label: "Embedded RDev Desktop".to_string(),
                kind: "screen".to_string(),
                backend: "rdev-desktop".to_string(),
                primary: true,
                ..Default::default()
            }],
            ..Default::default()
        };
    }

    #[cfg(not(feature = "embedded-rdev-desktop"))]
    {
        DesktopCapabilities {
            platform: std::env::consts::OS.to_string(),
            supported: false,
            view_only: true,
            input: false,
            clipboard: false,
            reason: "embedded GPU desktop service is not included in this build".to_string(),
            ..Default::default()
        }
    }
}

#[cfg(feature = "embedded-rdev-desktop")]
fn desktop_backends() -> Vec<String> {
    let mut backends = vec!["rdev-desktop".to_string(), "gpu-desktop-tunnel".to_string()];
    if cfg!(feature = "embedded-rdev-desktop-wayland") {
        backends.push("wayland".to_string());
        backends.push("pipewire".to_string());
    }
    if cfg!(feature = "embedded-rdev-desktop-vaapi") {
        backends.push("vaapi".to_string());
    }
    if cfg!(feature = "embedded-rdev-desktop-hw") {
        backends.push("nvenc".to_string());
    }
    backends
}

#[cfg(feature = "embedded-rdev-desktop")]
fn encoder_backends() -> Vec<String> {
    let mut encoders = vec!["libx264".to_string()];
    if cfg!(feature = "embedded-rdev-desktop-vaapi") {
        encoders.push("vaapi".to_string());
    }
    if cfg!(feature = "embedded-rdev-desktop-hw") {
        encoders.push("nvenc".to_string());
    }
    encoders
}

#[cfg(feature = "embedded-rdev-desktop")]
fn display_server() -> String {
    if std::env::var_os("WAYLAND_DISPLAY").is_some() {
        "wayland".to_string()
    } else if std::env::var_os("DISPLAY").is_some() {
        "x11".to_string()
    } else {
        String::new()
    }
}
